Marlin/MarlinSerial.cpp

```diff
@@ -310,3 +315,98 @@ MarlinSerial customizedSerial;
 #if defined(USBCON) && ENABLED(BLUETOOTH)
   HardwareSerial bluetoothSerial;
 #endif
+
+#if ENABLED(EMERGENCY_PARSER)
+
+  // Currently looking for: M108, M112, M410
+  // If you alter the parser please don't forget to update the capabilities in Conditionals.h
+
+  void emergency_parser(unsigned char c) {
+
+    enum e_parser_state {
+      state_RESET,
+      state_N,
+      state_M,
+      state_M1,
+      state_M10,
+      state_M108,
+      state_M11,
+      state_M112,
+      state_M4,
+      state_M41,
+      state_M410,
+      state_IGNORE // to '\n'
+    };
+
+    static e_parser_state state = state_RESET;
+
+    switch (state) {
+      case state_RESET:
+        switch (c) {
+          case ' ': break;
+          case 'N': state = state_N;      break;
+          case 'M': state = state_M;      break;
+          default: state = state_IGNORE;
+        }
+        break;
+
+      case state_N:
+        if (!NUMERIC_SIGNED(c) && c != ' ')
+          state = (c == 'M') ? state_M : state_RESET;
+        break;
```

Comment:
Sure enough, the `switch()` approach does generate smarter test ordering than the current logic.
```cpp
case state_N:
  switch (c) {
    case '0': case '1': case '2':
    case '3': case '4': case '5':
    case '6': case '7': case '8':
    case '9': case '-': case ' ':   break;
    case 'M': state = state_M;      break;
    default:  state = state_IGNORE;
  }
  break;
```
```asm
      case state_N:
        switch (c) {
    1ad8:	80 33       	cpi	r24, 0x30	; 48
    1ada:	08 f0       	brcs	.+2      	; 0x1ade <_Z16emergency_parserh+0x4c>
    1adc:	50 c0       	rjmp	.+160    	; 0x1b7e <_Z16emergency_parserh+0xec>
    1ade:	80 32       	cpi	r24, 0x20	; 32
    1ae0:	09 f4       	brne	.+2      	; 0x1ae4 <_Z16emergency_parserh+0x52>
    1ae2:	4d c0       	rjmp	.+154    	; 0x1b7e <_Z16emergency_parserh+0xec>
    1ae4:	8d 32       	cpi	r24, 0x2D	; 45
    1ae6:	61 f5       	brne	.+88     	; 0x1b40 <_Z16emergency_parserh+0xae>
    1ae8:	08 95       	ret
          default:  state = state_IGNORE;
        }
        break;
```